java/ql/lib/semmle/code/java/Expr.qll

```diff
@@ -161,18 +161,58 @@ class CompileTimeConstantExpr extends Expr {
     )
   }
 
+  /**
+   * Gets the stringified value of this expression, where possible.
+   *
+   * The stringified version of a compile-time constant expression is the equivalent to
+   * the result of calling `String.valueOf(expr)` on the expression.
+   * 
+   * Note that this does not handle the following cases:
+   *
+   * - mathematical computations of type `long`.
+   */
+  pragma[nomagic]
+  string getStringifiedValue() {
+    result = getStringValue()
+    or
+    result = this.(Literal).getValue()
+    or
+    result = this.getBooleanValue().toString()
+    or
+    result = this.getIntValue().toString()
+    or
+    // Ternary conditional, with compile-time constant condition.
+    exists(ConditionalExpr ce, boolean condition |
+      ce = this and
+      condition = ce.getCondition().(CompileTimeConstantExpr).getBooleanValue() and
+      result = ce.getBranchExpr(condition).(CompileTimeConstantExpr).getStringifiedValue()
+    )
+    or
+    exists(Variable v | this = v.getAnAccess() |
+      result = v.getInitializer().(CompileTimeConstantExpr).getStringifiedValue()
+    )
+  }
+
   /**
    * Gets the string value of this expression, where possible.
    */
   pragma[nomagic]
   string getStringValue() {
     result = this.(StringLiteral).getValue()
     or
-    result = this.(CharacterLiteral).getValue()
-    or
-    result =
-      this.(AddExpr).getLeftOperand().(CompileTimeConstantExpr).getStringValue() +
-        this.(AddExpr).getRightOperand().(CompileTimeConstantExpr).getStringValue()
+    (
+      if this.(AddExpr).getAnOperand().getType() instanceof TypeString // If either operand type is already a String:
+      then
+        // Then the stringified version of the expression can be safely used as a String will be created.
+        result =
+          this.(AddExpr).getLeftOperand().(CompileTimeConstantExpr).getStringifiedValue() +
+            this.(AddExpr).getRightOperand().(CompileTimeConstantExpr).getStringifiedValue()
+      else
+        // Adding two literals of primitive type will not result in a String.
+        result =
+          this.(AddExpr).getLeftOperand().(CompileTimeConstantExpr).getStringValue() +
+            this.(AddExpr).getRightOperand().(CompileTimeConstantExpr).getStringValue()
```

Comment:
```suggestion
      this.getType() instanceof TypeString and
      this.(AddExpr).getLeftOperand().(CompileTimeConstantExpr).getStringifiedValue() +
        this.(AddExpr).getRightOperand().(CompileTimeConstantExpr).getStringifiedValue()
```

Since if the result is a string, any non-string is getting `toString`'d, right?